tidy3d/plugins/design/method.py

```diff
@@ -23,335 +20,780 @@ class Method(Tidy3dBaseModel, ABC):
     name: str = pd.Field(None, title="Name", description="Optional name for the sweep method.")
 
     @abstractmethod
-    def run(self, parameters: Tuple[ParameterType, ...], fn: Callable) -> Tuple[Any]:
-        """Defines the search algorithm (sequential)."""
+    def _run(self, parameters: Tuple[ParameterType, ...], run_fn: Callable) -> Tuple[Any]:
+        """Defines the search algorithm."""
 
     @abstractmethod
-    def sample(self, parameters: Tuple[ParameterType, ...], **kwargs) -> Dict[str, Any]:
-        """Defines how the design parameters are sampled."""
+    def get_run_count(self, parameters: list = None) -> int:
+        """Return the maximum number of runs for the method based on current method arguments."""
+
+    def _force_int(self, next_point: dict, parameters: list) -> None:
+        """Convert a float asigned to an int parameter to be an int. Update dict in place."""
+
+        for param in parameters:
+            if isinstance(param, ParameterInt):
+                # Using int(round()) instead of just int as int always rounds down making upper bound value impossible
+                next_point[param.name] = int(round(next_point[param.name], 0))
 
     @staticmethod
-    def assert_hashable(fn_args: dict) -> None:
-        """Raise error if the function arguments aren't hashable (do before computation)."""
-        fn_args_tuple = tuple(fn_args)
-        try:
-            hash(fn_args_tuple)
-        except TypeError:
+    def _extract_output(output: list, sampler: bool = False) -> Tuple:
+        """Format the user function output for further optimization and result storage."""
+
+        # Light check if all the outputs are the same type
+        # If the user has supplied multiple returns this may catch it
+        # If the user has multiple returns and it passes this then it's less likely to cause further problems
+        if not all(isinstance(val, type(output[0])) for val in output):
             raise ValueError(
-                "Function arguments must be hashable. "
-                "Parameter sweep tool won't work with sets of lists, dicts or numpy arrays. "
-                "Convert these to 'tuple' for a workaround."
+                "Unrecognized output from supplied post function. The type of output varies across the output."
+                "Use of multiple return functions in fn_post is discouraged."
```

Comment:
double check any instances of python code / variable names in the exceptions and put single quotes `'` (`'fn_post'`) here

Reply:
What about pydantic field names? Is it ``stop_criteria_type`` or 'stop_criteria_type'?

Reply:
anything code related that's getting printed to terminal including field names should be in '. If it's going in docs (eg a docstring) it should be inside double `